### ISSUE TITLE
Warn about cases where a std::move is just a cast and no move is performed

### ISSUE THREAD

**Issue by firewave** [llvm/llvm-project#53083]
```cpp
#include <iostream>

class Move
{
public:
	Move() = default;

	Move(const Move& other) = delete;

	Move(Move&& other)
			: s_(std::move(other.s_))
	{
		std::cout << "move" << std::endl;
	}

	const std::string& s() {
		return s_;
	}

private:
	std::string s_{"str"};
};

static void f_rvalue(Move&& m)
{
	std::cout << "f_rvalue - " << m.s() << std::endl;
}

static void f_copy(Move m)
{
	std::cout << "f_copy - " << m.s() << std::endl;
}

static void f()
{
	Move m;
	f_rvalue(std::move(m)); // no move - cast only
	std::cout << "m.s() - " << m.s() << std::endl; // clang-tidy complains about moved variable being used

	Move m2;
	f_copy(std::move(m2)); // actual move happening
	std::cout << "m2.s() - " << m2.s() << std::endl; // clang-tidy complains about moved variable being used
}

int main()
{
	f();
	return 0;
}
```

Above code will output the following:
```
f_rvalue - str
m.s() - str
move
f_copy - str
m2.s() -
```

This is just a sample highlighting the "problem". I expected clang-tidy to warn about the non-working move similar to when you try to call a function with a const reference.
I am (now) aware that it is working as expected but clang-tidy is complaining about a moved object being used after the function call even though there is no move happening which is quite confusing - and possibly a false positive.

The issue in the original code was the different lifetime of the object/member since some consumers of it used the rvalue parameter. It was expected that the member is gone after the function call but for those cases it sticks around until the scope ends since the move never happens.

**Comment by dwblaikie** [comment]
I don't think this is something that'd be feasible to warn on in the compiler (though may be possible in the Clang Static Analyzer). Clang warnings don't do interprocedural analysis (so from the implementation of `f_rvalue` it can't see the `std::move` call to warn on that, and from the `std::move` call it can't see whether `f_rvalue` will move from it or not.

I think the warning's probably about right, though - if the function takes by rvalue that's a reasonable signal that the function /might/ move from the object. So the caller may need to only interact with the object conditionally - based on a return value from the function or the like. (not sure if the warning goes away if you do have such conditional code - I guess probably not... so /maybe/ that should be fixed)

**Comment by firewave** [comment]
> I don't think this is something that'd be feasible to warn on in the compiler (though may be possible in the Clang Static Analyzer).

I never expected this to become a compiler warning but something like a clang-tidy check.

> I think the warning's probably about right, though - if the function takes by rvalue that's a reasonable signal that the function /might/ move from the object.

Based on the fact that it took us quite a while to figure out what is going wrong here and that there are several blog posts on the net about this behavior this appears to be a rather common programming error. Though the clang-tidy warning shown in the sample is reasonable it does not reflect what actually happens and is thus rather misleading.

**Comment by JVApen** [comment]
Assuming this warning would get implemented the way you suggest:
 - How will you fix the code?
 - What will you do with forwarding references? 
 - What would be the behavior if you can't see the function implementation?
 - How would you deal with move constructors?

If I consider all that, I doubt a warning on called side is actually useful. What might make more sense is having the warning on implementation side. Aka: a function that takes an rvalue as an argument and doesn't move/use it in a way that the rvalue makes sense. As such, the function could be adapted to either take a (lvalue) reference or a by-value argument. 

That way, callers know what the expected behavior is: the thing is gone and left in a valid but unspecified state.

**Comment by firewave** [comment]
> Assuming this warning would get implemented the way you suggest:

I have no idea how to implement this. I just stated the most obvious warning you could give.

> * How will you fix the code?

In this case the fix would not be necessary on the callers side but on the consumer site. But what about cases where you have multiple signatures on the consumer side. In my sample I used different function names but what if they have the same name and just different ones. And there's also the const reference case. If you are not 100% familiar with move yet you could code weird things (I am talking from experience).

So this needs to be dealt with in an inverted way that you point out this in the implementation. But fixing that might cause existing code to break or generate follow-up warnings. I am not sure that is the way clang-tidy or static analyzer warnings are supposed to behave.

From my limited experience with move it seems like using `&&` in a regular function is not a good idea. You want to signal that the object should also be moved but than you end up in this pitfall. And if you don't you use it would get compiler or analyzer warnings in case it is copied so you catch these occurrences so you don't have to do it.

> * What will you do with forwarding references?

I have not really understood `std::forward` yet and only encountered it in code I did not write or when clang-tidy suggested I use it. And IIRC I can across a case where I had to fix the latter in some proof-of-concept code and that actually broke the code. So I have no input on that.

> * What would be the behavior if you can't see the function implementation?

That's obviously problematic. More reason for the check to be based on the implementation.

> * How would you deal with move constructors?

These probably need to be exclude or it could be a different check which makes sure the copy constructor always consumes the thing.

> If I consider all that, I doubt a warning on called side is actually useful. What might make more sense is having the warning on implementation side. Aka: a function that takes an rvalue as an argument and doesn't move/use it in a way that the rvalue makes sense. As such, the function could be adapted to either take a (lvalue) reference or a by-value argument.

You are absolutely write.
 
> That way, callers know what the expected behavior is: the thing is gone and left in a valid but unspecified state.

That's exactly what I want to achieve with the resulting code after you have addressed all findings.

**Comment by firewave** [comment]
Since you brought up that the suggested warning does not make sense on the calling side something a bit off-topic.

Looking at `performance-move-const-arg` which is reported on the caller side and offers a fix-it to remove the `std::move()`. This fix-it is incorrect if the intent was to move the object and the implementation is wrong. So that also raises the question on how to fix that. But since there is a fix-it it is always assumed that the caller is wrong and not the implementation. Maybe there should be no fix-it at all.
You should not blindly apply changes to your code but I think people consider fix-its as "safe" changes and just second guess them if they break the compilation.

**Comment by JVApen** [comment]
Let's try to keep this report focused, otherwise the person picking this to resolve is confused. If you believe that other check is faulty, let's log a separate bug for that.

**Comment by firewave** [comment]
> Let's try to keep this report focused, otherwise the person picking this to resolve is confused. If you believe that other check is faulty, let's log a separate bug for that.

I just wanted to point out that depending on the circumstances which caused such a warning the fix might differ. I filed another ticket for the case I brought up.

**Comment by firewave** [comment]
Based on https://github.com/llvm/llvm-project/issues/57266 it seems the information to provide such warnings is already available in some cases within clang-tidy.
